src/hotspot/share/opto/escape.cpp

```diff
@@ -95,15 +101,31 @@ void ConnectionGraph::do_analysis(Compile *C, PhaseIterGVN *igvn) {
   Compile::TracePhase tp("escapeAnalysis", &Phase::timers[Phase::_t_escapeAnalysis]);
   ResourceMark rm;
 
+#ifndef PRODUCT
+  elapsedTimer et;
+  et.start();
+#endif
   // Add ConP#NULL and ConN#NULL nodes before ConnectionGraph construction
   // to create space for them in ConnectionGraph::_nodes[].
   Node* oop_null = igvn->zerocon(T_OBJECT);
   Node* noop_null = igvn->zerocon(T_NARROWOOP);
   int invocation = 0;
   if (C->congraph() != NULL) {
     invocation = C->congraph()->_invocation + 1;
+#ifndef PRODUCT
+    // Reset counters when do_analysis is called again so objects are not double counted
```

Comment:
This does not look right either.  3 atomic words can't give you a safe transaction. 
I understand that you want to dedup across Iterative EAs. Snapshot is certainly a general solution but more complex. 

In our case,  I feel all we need is to compensate those java objects which have been scalarized in previous iterations. One member data _prev_scalar_replaced, which remembers the number of scalarized objects, seems good enough.   you can carry over this variable from the old to the new ConnectionGraph and add it back when the EA iterations end.

Reply:
@navyxliu Again, to avoid this I added local counters to the compiler that reset with every do_analysis call and once EA is completed and the while loop is executed they are atomically added to the static counters.